clang/test/SemaSYCL/constexpr-function-pointer.cpp

```diff
@@ -0,0 +1,41 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify -sycl-std=2020 -std=c++17 %s
+
+[[intel::device_indirectly_callable]] void t() {}
+
+const auto F = t;
+
+typedef void (*SomeFunc)();
+
+constexpr SomeFunc foo() { return t; }
+
+const SomeFunc foo1() { return t; }
+
+void bar1(const SomeFunc fptr) {
+  fptr();
+}
+
+template <auto f> void fooNTTP() { f(); }
+
+__attribute__((sycl_device)) void bar() {
+  // OK
+  constexpr auto f = t;
+  f();
+  // OK
+  const auto f1 = t;
+  f1();
```

Comment:
Is this not diagnosing because of `[[intel::device_indirectly_callable]]`? Can you briefly explain what the purpose of this attribute is (for my learning. I'm a bit confused)?

Reply:
Here is my understanding of `[[intel::device_indirectly_callable]]` (because I haven't seen the spec, but maybe @AlexeySachkov has):
It indicates that this function can be called indirectly inside the device code, so this attribute forces its availability inside the device code. In other words, even it is not called and doesn't have `SYCL_EXTERNAL` macro on it, it is still be emitted in the device code. Also it makes sure that function is externally visible and not optimized out, so at the same point it can be called indirectly. There is a doc for this attribute https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/AttrDocs.td#L3364 , though I'm not sure how outdated that is, at least I've never seen stuff like `cl::sycl::intel::get_device_func_ptr` .
Call to `f1` is not diagnosed because this patch. `f1` is a constant expression, so call through it will be lowered to direct call to `t` by CodeGen. So, the diagnostic is skipped. See variable `f2` - it is not a constant expression, so call to it is still diagnosed, because CodeGen won't make a direct call out of it.
I probably shouldn't have applied `[[intel::device_indirectly_callable]]` to function `t` because the test doesn't pass `-fsycl-allow-func-ptr` flag, so there is no any effects added by this attribute. Probably I worked too much on #5151 :D .
BTW, if I did pass this flag, the test would work in completely other way: any indirect calls would be allowed, but address taking of functions without this attribute would be disallowed even if the function pointer is stored to a constexpr variable.

Reply:
>> `f1 `is a constant expression, so call through it will be lowered to direct call to t by CodeGen.

`f1 ` is declared as `const auto` . This does not necessarily mean that it is evaluated at compile time right? 
https://godbolt.org/z/sT415qx7d

@premanandrao @erichkeane

Reply:
Thats right, there is nothing in the language that requires it be evaluated at compile time.  Consider that this:

````void foo(int a, int b) {
const auto c = a + b;
}````

Is valid.  There is obviously no way to const-evaluate that.

That said, the CFE DOES attempt to do constant evaluation on const variables when possible.

Reply:
> Here is my understanding of [[intel::device_indirectly_callable]] (because I haven't seen the spec, but maybe @AlexeySachkov has):
It indicates that this function can be called indirectly inside the device code, so this attribute forces its availability inside the device code. In other words, even it is not called and doesn't have SYCL_EXTERNAL macro on it, it is still be emitted in the device code. Also it makes sure that function is externally visible and not optimized out, so at the same point it can be called indirectly. There is a doc for this attribute https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/AttrDocs.td#L3364

That is correct understanding. We can take an address of a function in translation units, where it is only declared and perform an indirect call, for example.

Reply:
Ok, I was probably confused by the fact that
```
const int a = 1;
```
is a constant expression by C++ rules, but it seems this thing works only for integral types. Though, I've checked that the diagnostic is aligned with the fact that an expression is evaluated in codegen (I re-used the same set of functionality that CodeGen uses to check if it can do that). Maybe I should somehow rephrase the title of the commit? The thing is, together with this change some spec extension will be coming, I'm not sure that something like "can be evaluated at compile time" is clear and good thing when c++ actually has the idea of constexpr. Any suggestions?

Reply:
So that isn't a constant-expression by C++ rules, the language makes no requirement there as to the constant-expression-ness of that.  What differentiates a constant-expression is its usability in a constexpr or in a template argument.

CodeGen DOES use the constant-evaluator to try to save time by only emitting IR that we cannot already calculate, so that is probably what you're seeing. 

"Can be evaluated at compile time" isn't particularly useful either though, because the backend constant-folder can often solve things that we cannot in the CFE.

Reply:
Does it make sense to allow **only** things like this:
```
void t() {}
constexpr auto f = t; // I guess this time it is a constant expression?
f();
```
Because at least it can be normally specified.
Even though CFE can evaluate more cases.

Reply:
I think THAT is what I would go for, it actually fits properly in the rules of the language.  The constexpr keyword would give a meaningful error.  The CFE (and the backend!) can figure out even further, but we should do what is easily provable in Sema time.

A fun example:

```
typedef int (*Function)();
static Function Do;
static int EraseAll() { return system("rm -rf /"); }
void NeverCalled() { Do = EraseAll; }

int main() {
return Do();
}
```
This is a 'fun' case where the optimizer ends up deciding that "Do" MUST be 'EraseAll', so it changes main to be simply "call System("rm -rf /");" thanks to some fun UB.

See this article here for the explaination: https://kristerw.blogspot.com/2017/09/why-undefined-behavior-may-call-never.html

Reply:
Ok, done.

Reply:
> BTW, if I did pass this flag, the test would work in completely other way: any indirect calls would be allowed, but address taking of functions without this attribute would be disallowed even if the function pointer is stored to a constexpr variable.

@Fznamznon, is this still the case?   Can you add a variant to the current test case with the flag enabled, to compare what happens then?

Reply:
Passing the flag would require adding `[[intel::device_indirectly_callable]]` attribute to all functions of which we are taking address in this test I think. IMO, we should just remove the attribute from `void t(){}` (and not pass the flag) since it doesn't really have anything to do with this PR.

Reply:
Because the error is guarded by the flag so as to not break existing user code which may take address of function but not use it. More details are there here - https://github.com/intel/llvm/pull/5151 and https://github.com/intel/llvm/pull/5151#discussion_r783209864

Reply:
I agree with @elizabethandrews , SemaSYCL :: wrong-address-taking.cpp shows how diagnostics are working with the flag and the attribute just fine. I'll just remove the attribute.

Reply:
@Fznamznon, if I am looking at the right comment that @elizabethandrews is pointing to, you say this:

> Having in mind that the status of constexpr function pointers is still unknown, we have decided to implement diagnostic only under the flag as it allows us:
> 
> not to break existing codes where function pointer is taken, but not used
> have a diagnostic where function pointer is taken in order to be actually used later
> The flag serves as indicator that function pointers are actually going to be used, because user explicitly opt-ins for extra features

I agree with what you say above.

However, the test added there for the attribute, does not have constexpr cases.  This PR addresses the constexpr case, so I just want to understand when the two things are together.

Does that make sense?  I don't want to hold up the checking in of this PR, but it feels to me like two related features are being checked in without their interaction being tested.

Reply:
>> but it feels to me like two related features are being checked in without their interaction being tested.
Can you explain why it is related? I don't know if I am misunderstanding something. I thought `constexpr `doesn't affect the diagnostic. I think the diagnostic is generated based on whether the attribute is present or not (if the flag is passed). 

Since there is onfusion about this case, maybe it does make sense adding a test. I would suggest just adding a constexpr case in SemaSYCL :: wrong-address-taking.cpp

Reply:
@Fznamznon, thanks for explaining.  This was exactly the test case I was looking for (and I misunderstood what you had meant earlier by opposite behavior), and I thought we did not have a test for it.

I am all set now.   Sorry for the delay.

Reply:
Thanks @premanandrao . I know that my explanations can be unclear sometimes. I'm working on that :)
@elizabethandrews , are your concerns resolved?

Reply:
Yes. Thank you!